src/mpid/ch4/shm/src/shm_p2p.h

```diff
@@ -207,29 +299,103 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_SHM_mpi_irecv(void *buf, MPI_Aint count, MPI_
                                                  int rank, int tag, MPIR_Comm * comm,
                                                  int context_offset, MPIR_Request ** request)
 {
-    int ret;
-
+    int mpi_errno = MPI_SUCCESS;
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_SHM_MPI_IRECV);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_SHM_MPI_IRECV);
 
-    ret = MPIDI_POSIX_mpi_irecv(buf, count, datatype, rank, tag, comm, context_offset, request);
+#if (MPIDI_SHM_PT2PT_PROT == MPIDI_SHM_PT2PT_MULTIMODS)
+    MPIR_Comm *root_comm = NULL;
+    MPIR_Request *unexp_req = NULL;
+    MPIR_Context_id_t context_id = comm->recvcontext_id + context_offset;
+
+    /* When matches with an unexpected receive, it first tries to receive as
+     * a SHM optimized message (e.g., XPMEM SEND LMT). If fails, then receives
+     * as CH4 am message. Note that we maintain SHM optimized message in the
+     * same unexpected|posted queues as that used by CH4 am messages in order
+     * to ensure ordering.
+     */
+
+    /* Try to match with an unexpected receive request */
+    root_comm = MPIDIG_context_id_to_comm(context_id);
+    unexp_req = MPIDIG_dequeue_unexp(rank, tag, context_id, &MPIDIG_COMM(root_comm, unexp_list));
+
+    if (unexp_req) {
+        bool recvd_flag = false;
+
+        *request = unexp_req;
+        /* Mark as DEQUEUED so progress engine can complete a matched BUSY
+         * rreq once all data arrived */
+        MPIDIG_REQUEST(unexp_req, req->status) |= MPIDIG_REQ_UNEXP_DQUED;
```

Comment:
I'm not sure, but it seems `MPIDIG_REQUEST(unexp_req, req->status) |= MPIDIG_REQ_IN_PROGRESS;`  should be added too.

Reply:
Good point. I think we need it as we do not want the already matched receive be cancelled. Will fix.

Reply:
Fixed. running jenkins test.